triggers/daml/Daml/Trigger/Internal.daml

```diff
@@ -104,23 +103,28 @@ applyTransaction : Transaction -> ACS -> ACS
 applyTransaction (Transaction _ _ evs) acs = foldl (flip applyEvent) acs evs
 
 runRule
-  : (Party -> ACS -> Time -> Map CommandId [Command] -> s -> TriggerA ())
-  -> Time
-  -> TriggerState s
-  -> (TriggerState s, [Commands])
-runRule rule time state =
-  let (_, aState) =
-        runTriggerA
-          (rule state.party state.acs time state.commandsInFlight state.userState)
-          (TriggerAState state.commandsInFlight [] state.acs.pendingContracts state.nextCommandId)
-      commandsInFlight = foldl addCommands state.commandsInFlight aState.emittedCommands
-      acs = state.acs { pendingContracts = aState.pendingContracts }
-  in (state { nextCommandId = aState.nextCommandId, commandsInFlight, acs }, aState.emittedCommands)
-
-runTriggerA : TriggerA a -> TriggerAState -> (a, TriggerAState)
-runTriggerA (TriggerA f) s =
-  let (a, s') = runState f s
-  in (a, s' { emittedCommands = reverse s'.emittedCommands })
+  : (Party -> ACS -> Map CommandId [Command] -> s -> TriggerA ())
+  -> TriggerRule (TriggerState s) ()
+runRule rule = do
+  state <- get
+  cmds <- TriggerRule . zoom zoomIn zoomOut $ do
+    runTriggerRule . runTriggerA
+        $ rule state.party state.acs state.commandsInFlight state.userState
+    aState <- get
+    pure aState.emittedCommands
+  mapA_ submitCommands cmds -- TODO don't procrastinate
```

Comment:
I’m fine with leaving this for a separate PR if you want.

Reply:
Nah, we'll do it live. 0ea3d9c894